arrows/vxl/morphology.cxx

```diff
@@ -0,0 +1,325 @@
+{
+  auto input_plane = vil_plane( input, plane_index );
+  auto output_plane = vil_plane( output, plane_index );
+  switch( morphology_type )
+  {
+    case MORPHOLOGY_erode:
+    {
+      vil_binary_erode( input_plane, output_plane, morphological_element );
+      break;
+    }
+    case MORPHOLOGY_dilate:
+    {
+      vil_binary_dilate( input_plane, output_plane, morphological_element );
+      break;
+    }
+    case MORPHOLOGY_close:
+    {
+      vil_binary_closing( input_plane, output_plane, morphological_element );
+      break;
+    }
+    case MORPHOLOGY_open:
+    {
+      vil_binary_opening( input_plane, output_plane, morphological_element );
+      break;
+    }
+    case MORPHOLOGY_none:
+    {
+      output_plane.deep_copy( input_plane );
+      break;
+    }
+  }
+}
+
+// ----------------------------------------------------------------------------
+vil_image_view< bool >
+morphology::priv
+::perform_morphological_operations( vil_image_view< bool > const& input )
+{
+  setup_internals();
+
+  vil_image_view< bool > output{ input.ni(), input.nj(), input.nplanes() };
+
+  if( morphology_type != MORPHOLOGY_none )
+  {
+    for( auto i : vital::range::iota( input.nplanes() ) )
+    {
+      apply_morphology_to_plane( input, output, i );
+    }
+  }
+
+  if( combine_type == COMBINE_none )
+  {
+    // Don't combine across channels
+    return output;
+  }
+
+  // Select whether to do pixel-wise union or intersection
+  auto functor =
+    ( combine_type ==
+      COMBINE_union ) ? union_functor : intersection_functor;
+
+  auto accumulator = vil_plane( output, 0 );
+  for( unsigned i = 1; i < output.nplanes(); ++i )
+  {
+    auto current_plane = vil_plane( output, i );
+    // Union or intersect the current plane with the accumulator
+    vil_transform( accumulator, current_plane, accumulator, functor );
+  }
+  return accumulator;
+}
+
+// ----------------------------------------------------------------------------
+morphology
+::morphology()
+  : d{ new priv{} }
+{
+  attach_logger( "arrows.vxl.morphology" );
+}
+
+// ----------------------------------------------------------------------------
+morphology::~morphology() = default;
+
+// ----------------------------------------------------------------------------
+vital::config_block_sptr
+morphology
+::get_configuration() const
+{
+  // get base config from base class
+  vital::config_block_sptr config = algorithm::get_configuration();
+
+  config->set_value(
+    "morphology", morphology_converter().to_string( d->morphology_type ),
+    "Morphological operation to apply. Possible options are: " +
+    morphology_converter().element_name_string() );
+  config->set_value(
+    "element_shape", element_converter().to_string( d->element_type ),
+    "Shape of the structuring element. Possible options are: " +
+    element_converter().element_name_string() );
+  config->set_value( "element_size", d->element_size,
+                     "Size of the structuring element." );
```

Comment:
JFYI, this is a bit jargon-happy, though I'm not sure how this might be expressed instead. Maybe something about kernel size? If you have any thoughts, would be nice to reword this, but I wouldn't worry too much about it.